ui/src/shared/apis/queryCache.ts

```diff
@@ -0,0 +1,164 @@
+// Libraries
+import {sortBy} from 'lodash'
+
+// Utils
+import {asAssignment, getAllVariables} from 'src/variables/selectors'
+import {buildVarsOption} from 'src/variables/utils/buildVarsOption'
+import {filterUnusedVarsBasedOnQuery} from 'src/shared/utils/filterUnusedVars'
+import {runQuery} from 'src/shared/apis/query'
+
+// Types
+import {RunQueryResult} from 'src/shared/apis/query'
+import {CancelBox} from 'src/types/promises'
+import {AppState, Variable} from 'src/types'
+
+export const TIME_INVALIDATION = 5000
+
+const asSimplyKeyValueVariables = (vari: Variable) => {
+  return {
+    [vari.name]: vari.selected || [],
+  }
+}
+
+// Hashing function found here:
+// https://jsperf.com/hashcodelordvlad
+// Through this thread:
+// https://stackoverflow.com/questions/7616461/generate-a-hash-from-string-in-javascript
+const hashCode = (queryText: string): string => {
+  let hash = 0,
+    char
+  if (!queryText) {
+    return `${hash}`
+  }
+  for (let i = 0; i < queryText.length; i++) {
+    char = queryText.charCodeAt(i)
+    hash = (hash << 5) - hash + char
+    hash |= 0 // Convert to 32bit integer
+  }
+  return `${hash}`
+}
+
+class QueryCache {
+  cache = {}
+
+  _cleanExpiredQueries = (): void => {
+    const now = Date.now()
+    for (const id in this.cache) {
+      if (this.cache[id].isCustomTime) {
+        continue
+      }
+      if (now - this.cache[id].dateSet > TIME_INVALIDATION) {
+        this.resetCacheByID(id)
+      }
+    }
+  }
+
+  getFromCache = (id: string, variableID: string): RunQueryResult | null => {
+    // no existing query match
+    if (!this.cache[id]) {
+      return null
+    }
+    // query match with no existing variable match
+    if (this.cache[id].variables !== variableID) {
+      this.resetCacheByID(id)
+      return null
+    }
+    // query & variable match with an expired result
+    if (Date.now() - this.cache[id].dateSet > TIME_INVALIDATION) {
+      this.resetCacheByID(id)
+      return null
+    }
+    return this.cache[id].values
+  }
+
+  resetCacheByID = (id: string): void => {
+    if (!this.cache[id]) {
+      return
+    }
+    delete this.cache[id]
+  }
+
+  resetCache = (): void => {
+    this.cache = {}
+  }
+
+  setCacheByID = (
+    queryID: string,
+    variableID: string,
+    values: RunQueryResult,
+    isCustomTime: boolean = false
+  ): void => {
+    this.cache[queryID] = {
+      dateSet: Date.now(),
+      isCustomTime,
+      values,
+      variables: variableID,
+    }
+  }
+
+  startWatchDog = () => {
+    setInterval(() => {
+      this._cleanExpiredQueries()
+    }, TIME_INVALIDATION / 2)
+
+    this._cleanExpiredQueries()
+  }
+}
+
+const queryCache = new QueryCache()
+// Set an interval to check for expired data to invalidate
+queryCache.startWatchDog()
```

Comment:
so just by importing this file, we start a timer that runs every five seconds. in this PR i counted 5 imports of this file, not counting tests - does that mean five cache timers are going to be running every 5 seconds?

Reply:
Fixed this by only importing it once at the top of the file

Reply:
the original issue still remains i think? any place `apis/queryCache` is imported, a new `queryCache` will be instantiated and `startWatchDog` will be called, which starts the timers.

here's the issue with this: a person importing this file may not know that there are timers running when they import it. that may not be a problem, but in general, timers can be a big source of UI jank and frame drops, so we want to make sure we're not firing a bunch of them off needlessly.

i don't have a great handle on the architecture of this and it may be necessary for this to function correctly, it's just something to keep in the back of your mind in this time of UI performance

Reply:
Definitely a great point - and something I dreamt up some half-baked solutions to last night. As you mentioned, it is worth considering this point if/when we decide to expand the cache into other parts of the app. Something I was thinking about was to try and have the watchDog be context aware, but I'm not sure if that's even a thing? @drdelambre any thoughts?